### PR TITLE
Add paddle logical and reduce ops support.

### PR DESCRIPTION
**Pull request by huzhaoyX**
Signed-off-by: huzhaoyX <zhaoyangx.hu@intel.com>

### Details:
 - *Add paddle logical_and, logical_or and logical_xor ops support.*
 - *Add paddle reduce_max, reduce_mean, reduce_min, reduce_prod, and reduce_sum ops support.*

### CODE REVIEW

File: ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+template <typename T>
+NamedOutputs reduce_ops(const NodeContext& node) {
+    auto x = node.get_ng_input("X");
+    auto keep_dim = node.get_attribute<bool>("keep_dim");
+    auto reduce_all = node.get_attribute<bool>("reduce_all", false);
+    Shape input_shape = x.get_shape();
+    size_t input_rank = input_shape.size();
+    std::vector<int32_t> dims;
+    if (reduce_all) {
```
Comment (ceciliapeng2011) [inline ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp:21]: Do you get a test case with attr "reduce_all" is True?

File: ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+namespace frontend {
+namespace pdpd {
+namespace op {
+
+template <typename T>
+NamedOutputs reduce_ops(const NodeContext& node) {
+    auto x = node.get_ng_input("X");
+    auto keep_dim = node.get_attribute<bool>("keep_dim");
+    auto reduce_all = node.get_attribute<bool>("reduce_all", false);
+    Shape input_shape = x.get_shape();
```
Comment (ceciliapeng2011) [inline ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp:18]: X may have dynamic shape, which means we cannot get its shape here. Please refer to elementwise_ops.cpp get_partial_shape() about how to get the shape rank.

File: ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        }
+    } else {
+        dims = node.get_attribute<std::vector<int32_t>>("dim");
+    }
+    for (auto i : dims) {
+        std::ostringstream buf;
+        PDPD_OP_VALIDATION_CHECK(node, (i >= -(int)input_rank || i < (int)input_rank), buf.str());
+    }
+    std::vector<size_t> ng_reduction_axes_vect(dims.size());
+    std::transform(dims.begin(), dims.end(), ng_reduction_axes_vect.begin(), [input_rank](int idx) {
```
Comment (ceciliapeng2011) [inline ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp:33]: Why do we have to do this transform? axes can be negative, according to SPEC.

File: ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    size_t input_rank = input_shape.size();
+    std::vector<int32_t> dims;
+    if (reduce_all) {
+        for (int i = 0; i < (int)input_rank; ++i) {
+            dims.push_back(i);
+        }
+    } else {
+        dims = node.get_attribute<std::vector<int32_t>>("dim");
+    }
+    for (auto i : dims) {
```
Comment (ceciliapeng2011) [inline ngraph/frontend/paddlepaddle/src/op/reduce_ops.cpp:28]: Can you build a negative test case for it? If nGraph checks the violation, we can remove this check here.

File: ngraph/frontend/paddlepaddle/src/op/logical_ops.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+//
+
+#include <node_context.hpp>
+#include "default_opset.hpp"
+
+namespace ngraph {
+namespace frontend {
+namespace pdpd {
+namespace op {
+template <typename T>
```
Comment (ceciliapeng2011) [inline ngraph/frontend/paddlepaddle/src/op/logical_ops.cpp:12]: again, we can put all logical ops together. Better merge logical_not.cpp.

**Review by ceciliapeng2011 — CHANGES_REQUESTED** [review]

**Review by ceciliapeng2011 — APPROVED** [review]

**Comment by ilyachur** [comment]
Please rebase your branch to the latest master

**Comment by ilya-lavrenov** [comment]
@mvafin could you please review?

**Review by mvafin — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/src/op/logical_ops.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+NamedOutputs logical_ops(const NodeContext& node) {
+    auto x = node.get_ng_input("X");
+    auto y = node.get_ng_input("Y");
+    return node.default_single_output_mapping({std::make_shared<T>(x, y)}, {"Out"});
+}
+
+NamedOutputs logical_and(const NodeContext& node_context) {
```
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/src/op/logical_ops.cpp:20]: Can we use separate file for each conversion? As we have in the nGraph for operations?

Comment (ceciliapeng2011) [reply ngraph/frontend/paddlepaddle/src/op/logical_ops.cpp:20]: Sorry @ilyachur, not sure why you think we need separate them to multiple files. Each logical op is an instance of the template, simple and straightforward. There is a similar situation for elementwise ops already. And there are also multiple reduce_XX ops as well. I prefer to clustering close ops in a single file. How do you think?

**Comment by ilya-lavrenov** [comment]
@huzhaoyX can we proceed with this change?

**Comment by ilya-lavrenov** [comment]
@ceciliapeng2011 @meiyang-intel is it still valid?

**Comment by akladiev** [comment]
This PR will be closed in 2 weeks in case of no activity.

**Comment by akladiev** [comment]
This PR was closed because it has been stalled for 2 week with no activity.

**Pull request closed without merging**